### PR TITLE
blastem: init at 0.6.2-unstable-2024-03-31

### LINKED ISSUE #301134: Package request: blastem
**Project description**

BlastEm is an open source, highly accurate emulator for the Genesis/Megadrive.

Nightly release should be packaged as stable is five years out of date.

**Metadata**

* homepage URL: https://www.retrodev.com/blastem/
* source URL: https://www.retrodev.com/repos/blastem
* license: GPLv3
* platforms: unix

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/301134
@keenanweaver

---
This PR adds 1 package: `blastem`
`libretro.blastem` already exists, this is the original.

I put `blastem`, `dis`, `zdis` and `termhelper` into `$out/share/blastem`, but only `blastem` seem to do anything other than segfaulting or crashing. (I have no idea what they're supposed to do). Because of this I only made a wrapper for `blastem` in `$out/bin`.

I can't really test this ATM.

There was also a `menu.bin` file in the binary distribution of the package. I couldn't get it to build, because it would need `vasmm68k_mot`, which is not package in nixpkgs. Not sure if it's something important.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by keenanweaver** [comment]
Result of `nixpkgs-review pr 301312` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>blastem</li>
  </ul>
</details>

**Review by keenanweaver — APPROVED** [review]
Building & working perfectly. Thank you!

File: pkgs/by-name/bl/blastem/package.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    makeWrapper $out/share/blastem/blastem $out/bin/blastem
+
+    runHook postBuild
+  '';
+
+  meta = {
+    description = "The fast and accurate Genesis emulator";
+    homepage = "https://www.retrodev.com/blastem/";
+    license = lib.licenses.gpl3Plus;
+    maintainers = with lib.maintainers; [ tomasajt ];
```
Comment (wegank) [inline pkgs/by-name/bl/blastem/package.nix:57]: `meta.platforms` is not set.

https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#new-packages

Comment (TomaSajt) [reply pkgs/by-name/bl/blastem/package.nix:59]: Fixed :)

File: pkgs/by-name/bl/blastem/package.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  pname = "blastem";
+  version = "0.6.2-unstable-2024-03-31";
+
+  src = fetchhg {
+    url = "https://www.retrodev.com/repos/blastem";
+    rev = "48ab1e3e5df5";
+    hash = "sha256-UZl5fIE7LJqxwS8kFJ3xr8BJyHF60dnRNeA5k7lAuxg=";
+  };
+
+  nativeBuildInputs = [
```
Comment (Aleksanaa) [inline pkgs/by-name/bl/blastem/package.nix:22]: No `wrapGAppsHook` needed?

File: pkgs/by-name/bl/blastem/package.nix [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    runHook postBuild
+  '';
+
+  meta = {
+    description = "The fast and accurate Genesis emulator";
+    homepage = "https://www.retrodev.com/blastem/";
+    license = lib.licenses.gpl3Plus;
+    maintainers = with lib.maintainers; [ tomasajt ];
+    platforms = lib.platforms.all;
```
Comment (Aleksanaa) [inline pkgs/by-name/bl/blastem/package.nix:60]: Please add mainProgram.

File: pkgs/by-name/bl/blastem/package.nix [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  # Note: menu.bin cannot be generated yet, because it would
+  # need the `vasmm68k_mot` executable (part of vbcc for amigaos68k
+  # Luckily, menu.bin doesn't need to be present for the emulator to function
+
+  makeFlags = [ "HOST_ZLIB=1" ];
+
+  env.NIX_CFLAGS_COMPILE = "-I${lib.getDev SDL2}/include/SDL2";
+
+  installPhase = ''
+    runHook preBuild
```
Comment (Aleksanaa) [inline pkgs/by-name/bl/blastem/package.nix:42]:
```suggestion
    runHook preInstall
```

Also `postInstall`

**Comment by keenanweaver** [comment]
Result of `nixpkgs-review pr 301312` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>blastem</li>
  </ul>
</details>

**Review by keenanweaver — APPROVED** [review]

**Comment by Aleksanaa** [comment]
Doesn't build on aarch64-linux:
```
> Makefile:260: *** aarch64 is not a supported architecture.  Stop.
```

See https://github.com/NixOS/nixpkgs/pull/301312/checks?check_run_id=23660861694

**Comment by TomaSajt** [comment]
@ofborg build blastem

**Comment by TomaSajt** [comment]
@ofborg build blastem

**Comment by TomaSajt** [comment]
@ofborg build blastem

**Comment by TomaSajt** [comment]
@ofborg build blastem

**Comment by TomaSajt** [comment]
@ofborg build blastem

**Comment by TomaSajt** [comment]
@ofborg build blastem

**Comment by TomaSajt** [comment]
@ofborg build blastem

**Comment by TomaSajt** [comment]
@ofborg build blastem

**Comment by TomaSajt** [comment]
@ofborg eval

**Comment by TomaSajt** [comment]
@ofborg test blastem.passthru.version

**Comment by TomaSajt** [comment]
@ofborg build blastem.passthru.tests.version

**Comment by TomaSajt** [comment]
@ofborg build blastem.passthru.tests.version

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
> Doesn't build on aarch64-linux:
> 
> ```
> > Makefile:260: *** aarch64 is not a supported architecture.  Stop.
> ```
> 
> See https://github.com/NixOS/nixpkgs/pull/301312/checks?check_run_id=23660861694

This is ok, blastem is i386/x86_64 only: https://www.retrodev.com/blastem/.

**Pull request merged**